xpublish/utils/zarr.py

```diff
@@ -37,15 +36,40 @@
 logger = logging.getLogger('api')
 
 
-def _extract_dataset_zattrs(dataset: xr.Dataset) -> dict:
-    """Helper function to create zattrs dictionary from Dataset global attrs.
+def get_zvariables(dataset: xr.Dataset, cache: cachey.Cache):
+    """Returns a dictionary of zarr encoded variables, using the cache when possible."""
+    cache_key = dataset.attrs.get(DATASET_ID_ATTR_KEY, '') + '/' + 'zvariables'
+    zvariables = cache.get(cache_key)
 
-    Args:
-        dataset: The Dataset to extract zattrs from.
+    if zvariables is None:
+        zvariables = create_zvariables(dataset)
 
-    Returns:
-        A dictionary of zattrs.
-    """
+        # we want to permanently cache this: set high cost value
+        cache.put(cache_key, zvariables, 99999)
+
+    return zvariables
+
+
+def get_zmetadata(
+    dataset: xr.Dataset,
+    cache: cachey.Cache,
+    zvariables: dict,
+):
+    """Returns a consolidated zmetadata dictionary, using the cache when possible."""
+    cache_key = dataset.attrs.get(DATASET_ID_ATTR_KEY, '') + '/' + ZARR_METADATA_KEY
+    zmeta = cache.get(cache_key)
```

Comment:
This may be an issue for nested datasets, have to verify

Reply:
```
DATASET_ID_ATTR_KEY = '_xpublish_id'
```

Ok I think it just needs to be documented that if you build a dataset provider you should provide `_xpublish_id` as an attribute on the dataset or else the zarr attrs will be cached for the remote zarr dataset. Alternative is to change the dataset provider hook to assign an ID if it does not exist

Reply:
That makes sense, as I definitely think I've got dataset provider implementations that would have problems with this.